### PR TITLE
Piecemeal adoption of future language improvements

### PR DESCRIPTION
**Pull request by DougGregor**

### CODE REVIEW

File: proposals/nnnn-piecemeal-future-features.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+```
+
+The `hasFeature(X)` check indicates the presence of features, but by itself an older compiler will still attempt to parse the `#if` branch even if the feature isn't known. That's fine for this feature (implicitly opened existentials) because it doesn't add any syntax, but other features that add syntax might require something more. `hasFeature` can be composed with the `compiler` directive introduced by [SE-0212](https://github.com/apple/swift-evolution/blob/main/proposals/0212-compiler-version-directive.md), e.g.,
+
+```swift
+#if compiler(>=5.7) && hasFeature(BareSlashRegexLiterals)
+let regex = /.../
+#else
+let regex = NSRegularExpression("...")
+#endif
```
Comment (benrimmington) [inline proposals/nnnn-piecemeal-future-features.md:87]: Add `try` and `pattern:`.

```suggestion
#if compiler(>=5.7) && hasFeature(BareSlashRegexLiterals)
let regex = /.../
#else
let regex = try NSRegularExpression(pattern: "...")
#endif
```

File: proposals/nnnn-piecemeal-future-features.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+```swift
+#if compiler(>=5.7)
+#if hasFeature(BareSlashRegexLiterals)
+let regex = /.../
+#else
+let regex = NSRegularExpression("...")
+#else
+let regex = NSRegularExpression("...")
+#endif
```
Comment (benrimmington) [inline proposals/nnnn-piecemeal-future-features.md:100]: Add `#/.../#`, `#endif`, `try`, and `pattern:`.

```suggestion
#if compiler(>=5.7)
  #if hasFeature(BareSlashRegexLiterals)
  let regex = /.../
  #else
  let regex = #/.../#
  #endif
#else
let regex = try NSRegularExpression(pattern: "...")
#endif
```

File: proposals/nnnn-piecemeal-future-features.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+This proposals explicitly embraces the piecemeal, intentional adoption of features that were held until Swift 6 for source-compatibility reasons. It establishes a direct path to incrementally adopt Swift 6 features, one-by-one, to gain their benefits in a Swift 4.x/5.x code base and smooth the migration path to a Swift 6 language mode. Developers can use a new compiler flag, `-enable-future-feature X` to enable the specific feature named `X` for that module, and multiple features can be specified in this manner. When the developer moves to the next major language version, `X` will be implied by that language version and the compiler flag will be rejected. This way, future feature flags only accumulate up to the next major Swift language version and are then cleared away, so we don't fork the language into incompatible dialects.
```
Comment (WowbaggersLiquidLunch) [inline proposals/nnnn-piecemeal-future-features.md:26]: typo: "This proposals" → "This proposal"

**Comment by stevapple** [comment]
It would be better to mention the impact on Swift Evolution — we can now raise and accept source-breaking proposals, and each accepted proposal that contains a "future" feature should explicitly mark its feature name and first available version (with unified format?), instead of just `Implemented (Swift Next)` or `Partially implemented on main` or `available behind … flags`.

**Comment by hborla** [comment]
The review is up! https://forums.swift.org/t/se-0362-piecemeal-adoption-of-future-language-improvements/58384

I agree that the impact on the Swift Evolution process should be discussed on the review thread.

**Comment by stevapple** [comment]
> That said, can we discuss this on the review thread once the review manager opens it, rather than here on the pull request?

Ah of course!

I just wrote the lines even before this PR was merged:) It took me minutes and I didn’t notice that a review thread is immediately followed.

**Pull request merged**